src/sentry/sentry_metrics/querying/data/plan.py

```diff
@@ -59,11 +58,51 @@ class MetricsQueriesPlan:
     For example, you could define a simple query "a: max(mri_1)" and use it in the formula as "$a".
     """
 
+    queries: Mapping[str, str] | None = None
+    formulas: Sequence[FormulaDefinition] | None = None
+
+    def get_replaced_formulas(self) -> Sequence[FormulaDefinition]:
+        """
+        Returns a list of formulas with the variables replaced with the actual mql query string.
+
+        The usage of a variable in the formulas is with the `$` + the name of the query. The rationale
+        behind choosing `$` is to keep the syntax compatible with the MQL syntax, in case we were to embed
+        variables resolution in the layer itself.
+
+        This function naively uses string substitution to replace the contents. In case we see it's too
+        fragile, we might want to switch to parsing the actual input and mutating the AST.
+
+        Returns:
+            A list of FormulaDefinition objects that contain the final MQL to execute against Snuba.
+        """
+        if not self.queries:
+            return self.formulas
+
+        return list(map(lambda formula: formula.replace_variables(self.queries), self.formulas))
+
+    def is_empty(self) -> bool:
+        """
+        A query plan is defined to be empty is no formulas have been applied on it.
+
+        Returns:
+            A boolean which is True when the plan is empty, or False otherwise.
+        """
+        return not self.formulas
+
+
+class MetricsQueriesPlanBuilder:
```

Comment:
Created a builder to build the `MetricsQueriesPlan` object since we want to give people the ability to build it without using the builder pattern.